samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -392,6 +395,90 @@ StartpointManager createStartpointManager() {
     return new StartpointManager(coordinatorStreamStore);
   }
 
+  /**
+   * Check if the new job model contains a different work assignment for the processor compared the last active job
+   * model. In case of different work assignment, expire the current job model by invoking the <i>onJobModelExpired</i>
+   * on the registered {@link JobCoordinatorListener}.
+   * At this phase, the job model is yet to be agreed by the quorum and hence, this optimization helps availability of
+   * the processors in the event no changes in the work assignment.
+   *
+   * @param newJobModel new job model published by the leader
+   */
+  @VisibleForTesting
+  void checkAndExpireJobModel(JobModel newJobModel) {
+    Preconditions.checkNotNull(newJobModel, "JobModel cannot be null");
+    if (coordinatorListener == null) {
+      LOG.info("Skipping job model expiration since there are no active listeners");
+      return;
+    }
+
+    if (JobModelUtil.compareContainerModelForProcessor(processorId, activeJobModel, newJobModel)) {
+      LOG.info("Skipping job model expiration for processor {} due to no change in work assignment.", processorId);
```

Comment:
Minor:
Would be helpful to either log the jobModel or the work-assignment here for debugging purposes. Essentially logging the before/after work-assignment on a quorum change just for debugging purposes. May-be logging it when they're different would be better.